scripts/builtin/hmm.dml

```diff
@@ -0,0 +1,258 @@
+    A = rand(rows=nr_states, cols=nr_states)
+    T = ncol(eta)
+
+    parfor (i in 1:nr_states) {
+        parfor (j in 1:nr_states) {
+            eta_id = (i-1) * nr_states + j
+            num_ij = sum(eta[eta_id, 1:T-1])
+            den_ij = sum(gamma[i,1:T-1])
+            A[i,j] = num_ij/den_ij
+        }
+    }
+}
+
+calculate_B = function(Matrix[Double] gamma, Matrix[Double] X, Integer nr_states, Integer nr_outputs) return (Matrix[Double] B) {
+    B = rand(rows=nr_states, cols=nr_outputs)
+
+    parfor (i in 1:nr_states){
+        parfor (j in 1:nr_outputs) {
+            B[i, j] = sum((X == j) * gamma[i, ]) / sum(gamma[i, ])
+        }
+    }
+}
+
+/*
+  To work with 3d matrices, the values in first two dimensions of eta matrix
+  must be rolled the first dimension. The following indexing function maps the
+  index of rolled matrix(eta) to "normal" 2d matrices like alpha, beta, etc
+*/
+index = function(Integer rolled_i, Integer nr_states) return(Double i, Double j) {
+    i = ceil(rolled_i / nr_states) #index starts at 1
+    j = rolled_i - (nr_states * (i-1)) #i-1) * 
+}
+
+calculate_eta =  function (Matrix[Double] X, Matrix[Double] alpha, Matrix[Double] beta, Matrix[Double] A, Matrix[Double] B) return (Matrix[Double] eta)
+{
+    nr_states = nrow(alpha)
+    T = ncol(alpha)    
+    
+    /*
+    eta a (nr_states * nr_states) * T matrix with cell (i, t) being probability of 
+    the state being at i at timestep j given the observed output
+    */
+    
+    tot_transitions = nr_states * nr_states
+    eta = rand(rows=tot_transitions, cols=T-1)
+    
+    /*
+    The transitions will be indiced as such:transition 1-N will represent
+    transition from state 1 to 1, 2 upto state N. transition N+1-2N will represent
+    transitions from 2 to 1, 2 upto state N  
+    */
+    
+    parfor (trans_id in 1:tot_transitions) {
+        [i, j] = index(trans_id, nr_states)
+        for (t in 1:(T-1)) {
+            #indices for alpha and beta
+            ot1 = output_t(X, t+1)
+            num_ij = alpha[i, t] * A[i, j] * beta[j, t+1] * B[j, ot1]
+            den = matrix(0, rows=nr_states, cols=1)
+            
+            parfor (k in 1:nr_states) {
+                den[k, 1] = sum(alpha[k, t] * t(A[k, ]) * beta[, t+1] * B[,ot1])
+            }
+            s = num_ij / sum(den)
+            eta[trans_id, t] = as.scalar(s[1,1])
+        }
+    }
+}
+
+calculate_gamma = function (Matrix[Double] alpha, Matrix[Double] beta) return (Matrix[Double] gamma)
+{ 
+    /*
+    gamma a nr_state * T matrix with cell (i, t) being probability of 
+    the state being at i at timestep j given the observed output 
+    */
+    nr_states = nrow(alpha)
+    T = ncol(alpha)
+    
+    gamma = rand(rows=nr_states, cols=T)
+    
+    parfor (i in 1:nrow(alpha)) {
+        parfor (t in 1:ncol(alpha)) {
+            num_it = alpha[i, t] * beta[i, t]
+            den_it = sum(alpha[,t] * beta[,t])
+            gamma[i, t] = num_it / den_it
+        }
+    }
```

Comment:
I think these parfors can be rewritten as:

```
num_it_M = alpha * beta
den_it_M =  rowsum(num_it_M)
gamma = num_it_M / den_it_M
```

This is because we support direct matrix and vector operations.
And i think this nicely maps to it. (But i might have an error here.)